pkg/auth/auth.go

```diff
@@ -0,0 +1,171 @@
+	resourceList, err := discoveryCli.ServerResourcesForGroupVersion(groupVersion)
+	if err != nil {
+		return "", nil
+	}
+	for _, r := range resourceList.APIResources {
+		if r.Kind == kind {
+			return r.Name, nil
+		}
+	}
+	return "", fmt.Errorf("Unable to find the kind %s in the resource group %s", kind, groupVersion)
+}
+
+func (u *UserAuth) canPerform(verb, group, resource, namespace string) (bool, error) {
+	res, err := u.authCli.SelfSubjectAccessReviews().Create(&authorizationapi.SelfSubjectAccessReview{
+		Spec: authorizationapi.SelfSubjectAccessReviewSpec{
+			ResourceAttributes: &authorizationapi.ResourceAttributes{
+				Group:     group,
+				Resource:  resource,
+				Verb:      verb,
+				Namespace: namespace,
+			},
+		},
+	})
+	if err != nil {
+		return false, err
+	}
+	return res.Status.Allowed, nil
+}
+
+func (u *UserAuth) getResourcesToCheck(namespace, manifest string) ([]resource, error) {
+	objs, err := yamlUtils.ParseObjects(manifest)
+	if err != nil {
+		return []resource{}, err
+	}
+	resourcesToCheck := map[string]*resource{}
+	result := []resource{}
+	for _, obj := range objs {
+		// Object can specify a different namespace, if not use the default one
+		ns := obj.GetNamespace()
+		if ns == "" {
+			ns = namespace
+		}
+		resourceToCheck := fmt.Sprintf("%s/%s/%s", ns, obj.GetAPIVersion(), obj.GetKind())
+		if resourcesToCheck[resourceToCheck] == nil {
+			r := resource{obj.GetAPIVersion(), obj.GetKind(), ns}
+			resourcesToCheck[resourceToCheck] = &r
+			result = append(result, r)
+		}
+	}
+	return result, nil
+}
+
+func (u *UserAuth) isAllowed(verb string, itemsToCheck []resource) error {
+	for _, i := range itemsToCheck {
+		resource, err := resolve(u.discoveryCli, i.APIVersion, i.Kind)
+		if err != nil {
+			return err
+		}
+		group := i.APIVersion
+		if group == "v1" {
+			// The group should be empty for the core API group
+			group = ""
+		}
+		allowed, _ := u.canPerform(verb, group, resource, i.Namespace)
+		if !allowed {
+			return fmt.Errorf("Unauthorized to %s %s/%s in the %s namespace", verb, i.APIVersion, resource, i.Namespace)
+		}
+	}
+	return nil
+}
+
+// CanI returns if the user can perform the given action with the given chart and parameters
+func (u *UserAuth) CanI(namespace, action, manifest string) error {
+	resources, err := u.getResourcesToCheck(namespace, manifest)
+	if err != nil {
+		return err
+	}
+	switch action {
+	case "create":
+		err := u.isAllowed("create", resources)
+		if err != nil {
+			return err
+		}
+	case "upgrade":
+		// For upgrading a chart the user should be able to create, update and delete resources
+		for _, v := range []string{"create", "update", "delete"} {
+			err = u.isAllowed(v, resources)
+			if err != nil {
+				return err
+			}
+		}
+	case "get":
+		err := u.isAllowed("get", resources)
+		if err != nil {
+			return err
+		}
+	case "delete":
+		err := u.isAllowed("delete", resources)
+		if err != nil {
+			return err
+		}
+	default:
```

Comment:
I think we can just have the default be `err := u.isAllowed(action, resources` and remove all other cases apart from `upgrade`. The SelfSubjectAccessReview will fail for unknown verbs anyway.